lib/Dialect/TTNN/Transforms/TTNNWorkarounds.cpp

```diff
@@ -0,0 +1,418 @@
+#include <llvm/ADT/SmallVector.h>
+#include <llvm/Support/Casting.h>
+#include <llvm/Support/LogicalResult.h>
+#include <llvm/Support/raw_ostream.h>
+#include <mlir/IR/BuiltinTypes.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/IR/Value.h>
+#include <mlir/IR/ValueRange.h>
+#include <mlir/Interfaces/DestinationStyleOpInterface.h>
+#include <mlir/Rewrite/FrozenRewritePatternSet.h>
+#include <mlir/Support/LLVM.h>
+#include <mlir/Transforms/GreedyPatternRewriteDriver.h>
+#include <vector>
+
+namespace mlir::tt::ttnn {
+#define GEN_PASS_DEF_TTNNWORKAROUNDS
+#include "ttmlir/Dialect/TTNN/Transforms/Passes.h.inc"
+
+// Helper method to apply tensor layout workaround. It accepts workaround and
+// current tensor layout as input arguments. It applies the workaround on a
+// current tensor layout and returns true if the tensor layout was modified.
+static bool applyTensorLayoutWorkaround(wa::TTNNOperandWorkarounds &workaround,
+                                        Layout &tensorLayout) {
+  bool modified = false;
+  if (workaround.getWorkaround<Layout>()) {
+    modified = tensorLayout != *workaround.getWorkaround<Layout>();
+    tensorLayout = *workaround.getWorkaround<Layout>();
+  }
+
+  return modified;
+}
+
+// Helper method to apply tensor buffer type workaround. It accepts workaround
+// and current tensor buffer type as input arguments. It applies the workaround
+// on a current tensor buffer type and returns true if the tensor buffer type
+// was modified.
+static bool
+applyTensorBufferTypeWorkaround(wa::TTNNOperandWorkarounds &workaround,
+                                BufferType &tensorBufferType) {
+  bool modified = false;
+  if (workaround.getWorkaround<BufferType>()) {
+    modified = tensorBufferType != *workaround.getWorkaround<BufferType>();
+    tensorBufferType = *workaround.getWorkaround<BufferType>();
+  }
+
+  return modified;
+}
+
+// Helper method to apply tensor memory layout workaround. It accepts workaround
+// and current tensor memory layout as input arguments. It applies the
+// workaround on a current tensor memory layout and returns true if the tensor
+// memory layout was modified.
+static bool
+applyTensorMemoryLayoutWorkaround(wa::TTNNOperandWorkarounds &workaround,
+                                  TensorMemoryLayout &tensorMemoryLayout) {
+  bool modified = false;
+  if (workaround.getWorkaround<TensorMemoryLayout>()) {
+    modified =
+        tensorMemoryLayout != *workaround.getWorkaround<TensorMemoryLayout>();
+    tensorMemoryLayout = *workaround.getWorkaround<TensorMemoryLayout>();
+  }
+
+  return modified;
+}
+
+// Helper method to propagate output changes into DPS inits. It accepts output
+// result as input argument. It iterates over all the uses of the output result
+// and updates the DPS init with the new output type.
+static void propagateOutputChangesIntoDPSInits(OpResult &outputResult) {
+  // Iterate over all the uses of the outputResult
+  for (auto &use : outputResult.getUses()) {
```

Comment:
Are we allowed to do this inside of a RewritePattern?

Reply:
All modifications must be performed within the rewriter context. Initially, I invoked the propagateOutputChangesIntoDPSInits method inside the rewriter, but I have updated the code to align better with this requirement, ensuring that all changes occur under the rewriter.